File: src/Main/150/235.cc
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */

class Solution {
  public:
    TreeNode *lowestCommonAncestor(TreeNode *root, TreeNode *p, TreeNode *q) {
        if (q->val < p->val) swap(p, q);
        return f(root, p, q);
    }

    TreeNode *f(TreeNode *root, TreeNode *p, TreeNode *q) {
        if (!root) return nullptr;
        if (root->val >= p->val && root->val <= q->val) {
            return root;
        } else if (root->val > q->val) {
            return f(root->left, p, q);
        } else {
            return f(root->right, p, q);
        }
    }
};

File: src/archive/Zon/AmazonPrepOld/Trees/go/235.go
package trees

// Definition for a binary tree node
type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func lowestCommonAncestor(root, p, q *TreeNode) *TreeNode {
	if q.Val < p.Val {
		temp := p
		p = q
		q = temp
	}

	return dfs(root, p, q)
}

func dfs(root, p, q *TreeNode) *TreeNode {
	if root == nil {
		return root
	}
	if p.Val <= root.Val && root.Val <= q.Val {
		return root
	} else if p.Val > root.Val {
		return dfs(root.Right, p, q)
	} else {
		return dfs(root.Left, p, q)
	}
}
